### PR TITLE
Libai bench merge main

### PR DESCRIPTION
**Pull request by CPFLAME**
按照 https://github.com/Oneflow-Inc/OneTeam/issues/1500 修正LiBai中的代码,

- [x] 添加issue中的提到的各个功能与优化
- [x] 正确性与性能验证

### CODE REVIEW

File: libai/data/structures.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ class DistTensorData:
     placement_idx: int = 0
 
     # Tensor-like methods
-    def to_global(self, sbp=None, placement=None):
+    def to_global(self, sbp=None, placement=None, device_type="cuda"):
```
Comment (strint) [inline libai/data/structures.py:32]: 一般涉及更换  device 的，应该用 to，看这里是不是区分下接口

Comment (CPFLAME) [reply libai/data/structures.py:32]: 在这里主要是为了兼容 两种策略

https://github.com/Oneflow-Inc/OneTeam/issues/1500#issuecomment-1161183526
和
https://github.com/Oneflow-Inc/OneTeam/issues/1500#issuecomment-1161192641

当用户关心流水并行的效率时,  直接用默认的"cuda"就好.  
但是如果用户想在流水并行时效率更高, 则需要在configs里面设置为"cpu", 然后还需要在model.forward()里面写"to_global()"

Comment (yuanms2) [reply libai/data/structures.py:32]: “当用户关心流水并行效率时”   和  "如果用户想在流水并行时效率更高“

感觉这俩时一个情况，但为什么分别是cuda和cpu

Comment (CPFLAME) [reply libai/data/structures.py:32]: 我写错了  
应该是 “当用户不关心流水并行效率时”,  也就是用户不用额外的在model.forward()函数里面写to_global()语句

**Comment by CPFLAME** [comment]
loss 已经对齐,

但是需要关闭 rdma和 注释掉 dataloader中的`persistent_workers=True if num_workers > 0 else False,`

看上去 `persistent_workers`这个参数会对dataloader的随机性造成影响?  我怀疑加上后 拿到的好像不是相同的数据


image

**Comment by yuanms2** [comment]
rdma 和 persistent worker 的影响分别是什么，是还有未合并的fix吗

**Comment by Flowingsun007** [comment]
> 看上去 `persistent_workers`这个参数会对dataloader的随机性造成影响?

persistent_workers这个参数确实会影响随机性，这个是其机制之一。因为persistent_workers=True时worker持久存在，所以随机种子都是用的一样的；设为False时，则不会共享

**Comment by CPFLAME** [comment]
本地跑`model_test.sh` 已经通过. 
该pr可以review与合并了

File: libai/models/utils/graph_base.py [hunk 2]
```diff
@@ -105,6 +106,17 @@ def build(self, **kwargs):
             loss_dict = self.model(**kwargs)
             losses = sum(loss_dict.values())
             losses.backward()
+            # set loss_dict on rank0
+            # Consider if it's 2d mesh, ranks should be [[0]] instead of [0]
```
[the remark this answers is not in the document]
Comment (L1aoXingyu) [reply libai/models/utils/graph_base.py:110]: 为了在 graph 内部做 boxing，之前是放在外面做的，所以会有 eager boxing 阻断流水

**Review by chengtbf — APPROVED** [review]

**Comment by chengtbf** [comment]
这个 PR 可以合并了。我没有问题了

File: libai/data/structures.py (hunk 1 above)
Comment (strint) [inline libai/data/structures.py:32]: 这里有假设原 self.tensor 是 local 还是 global 么？

这个定义的不是太清晰，和 oneflow 本身同名，但是行为又不一样。

看起来像是一个带状态的 global tensor 的 factory？
- 初始化一个配置
- 可以根据初始化配置 build
- 可以根据初始化配置 rebuild
- build 和 rebuild 的结果，还保存下来了，放到了 self.tesnor

当前使用可能没有正确性问题，但是后面最好重构下

**Review by strint — APPROVED** [review]

File: libai/data/structures.py (hunk 1 above)
Comment (CPFLAME) [reply libai/data/structures.py:32]: 这里既定的假设 self.tensor 是local的, 在dataset的返回里面直接包一层DistTensorData就可以了

原意是想不让用户接触 to_global, sbp, placement这些概念, 直接在LiBai里面调用这个接口就行, 不用搞懂里面的原理是什么, 所以封装了一下

File: libai/layers/attention.py [hunk 3]
```diff
@@ -204,22 +213,37 @@ def forward(
...
-
-        # [bsz, num_heads, tgt_len, src_len]
-        attention_weights = self.dropout(attention_weights)
+            if self.scale_mask_softmax_fusion and self.attn_mask_type == AttnMaskType.causal:
+                attention_weights = flow._C.fused_scale_tril_softmax_mask_scale(
+                    attention_scores,
+                    p=self.attention_dropout_prob,
+                    diagonal=0,
+                    tril_scale_value=self.coeff,
+                )[0]
```
Comment (CPFLAME) [inline libai/layers/attention.py:242]: fused_scale_tril_softmax_mask_scale 这个算子好像计算结果有点问题, 

和225行~234行代码得到的运算结果不一样.
```python
                if self.coeff is not None:
                    attention_scores *= self.coeff
                attention_scores = flow.mul(attention_scores, attention_mask)
                attention_scores = attention_scores - 10000.0 * (1 - attention_mask)
                # TODO(xingyu.liao): graph will occur `where_scalar` errors
                # when using `masked_fill`
                # attention_scores = attention_scores.masked_fill(1 - attention_mask, -10000.0)
                attention_weights = flow.softmax(attention_scores, dim=-1)
                # [bsz, num_heads, tgt_len, src_len]
                attention_weights = self.dropout(attention_weights)
```
